fcrepo/fcrepo
# Refactor referenceService to emit events.

Fix AbstractJmsIT to stop checking message headers.

Fix JsonLdEventMessage to have only one type for Object and Actor.


**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3434

# What does this Pull Request do?
Implements the INBOUND_REFERENCE event type when you reference an item.

Fixes the fcrepo-jms tests to stop relying on the message headers and actually decode the body of the message.

# How should this be tested?

1. Setup a JMSClient to listen to Fedora.
1. Create two resources.
1. Make one reference the other ie. `<http://localhost:8080/rest/first> <http://example.org/pointsTo> <http://localhost:8080/rest/second>`
1. Check your JMSClient for a Follow event, like
    ```
    {"id":"urn:uuid:8cea1bab-c1c1-4b9b-ac5d-a666aa6abad1","type":["Follow"],"name":"refer to resource","published":"2020-09-25T20:57:08.641850Z","actor":[{"type":"Person","id":"info:fedora/local-user#fedoraAdmin"},{"type":"Application","name":"curl/7.64.1"}],"object":{"type":["http://mementoweb.org/ns#TimeGate","http://fedora.info/definitions/v4/repository#Container","http://www.w3.org/ns/ldp#RDFSource","http://www.w3.org/ns/ldp#Resource","http://www.w3.org/ns/ldp#Container","http://mementoweb.org/ns#OriginalResource","http://fedora.info/definitions/v4/repository#Resource","http://www.w3.org/ns/ldp#BasicContainer","http://www.w3.org/ns/prov#Entity"],"id":"http://localhost:8080/rest/second","isPartOf":"http://localhost:8080/rest"},"@context":["https://www.w3.org/ns/activitystreams#",{"prov":"http://www.w3.org/ns/prov#","dcterms":"http://purl.org/dc/terms/","type":"@type","id":"@id","isPartOf":{"@id":"dcterms:isPartOf","@type":"@id"}}]}
    ```
1. Do the same but make the target of the reference a resource that does NOT exist.
1. See NO Follow event.


# Interested parties
@fcrepo4/committers
